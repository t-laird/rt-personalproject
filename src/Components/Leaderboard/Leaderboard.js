import React, { Component } from 'react';
import { connect } from 'react-redux';
import LeaderboardDisplay from '../LeaderboardDisplay/LeaderboardDisplay';
import './Leaderboard.css';
import PropTypes from 'prop-types';

export class Leaderboard extends Component {

  getData = (value) => {
    const recentWeek = this.props.GroupTransactions[this.props.GroupTransactions.length - 1].transactions;
    const givers = recentWeek.reduce((accum, transaction) => {
      if (!accum[transaction[value]]) {
        accum[transaction[value]] = { points: 0 };
      }
      accum[transaction[value]].points += transaction.point_value;
      return accum;
    }, {});
    const array = Object.keys(givers).map(user => {
      return { name: user, points: givers[user].points };
    });
    const result = array.sort((a, b) => b.points - a.points);

    return result.slice(0, 5);
  }

  displayLeaders = (dataSet) => {
    if (this.props.GroupTransactions.length) {
      const data = this.getData(dataSet);
      const display = data.map((obj, index) => {
        return <LeaderboardDisplay 
          name={obj.name}
          points={obj.points}
          key={index}
          position={index + 1}
        />;
      });

      return display;
    }
  }

  render() {
    return (
      <div className="leaderboard-component">
        <div className="givers">
          <h2>Top Givers this week:</h2>
          {this.displayLeaders('send_name')}
        </div>
        <div className="senders">
          <h2>Top Earners this week:</h2>
          {this.displayLeaders('received_name')}
        </div>
      </div>
    );
  }
}

export const mapStateToProps = ( store ) => ({
  GroupTransactions: store.GroupTransactions
});

export default connect(mapStateToProps, null)(Leaderboard);

Leaderboard.propTypes = {
  GroupTransactions: PropTypes.array
};