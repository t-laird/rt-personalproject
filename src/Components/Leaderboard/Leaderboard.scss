@import '../../Variables.scss';

.leaderboard-component {
  display: flex;
  justify-content: space-between;
  width: 650px;
  margin: 0 auto 40px;

  h2, h3 {
    color: #fff;
  }

  h2 {
    border-bottom: 6px solid rgba(0,0,0, .2);
    margin-bottom: 8px;
    padding-bottom: 7px;
    width: 305px;
    text-align: center;
  }
}