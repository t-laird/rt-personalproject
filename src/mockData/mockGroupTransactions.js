const mockGroupTransactions = [
  {transactions: [
    {
      created_time: 1514939019586,
      event_id: 9,
      event_time: "2018-01-03T00:23:39.627Z",
      group_id: 1,
      note: "You totally killed it on your lightning talk this morning!  I loved it.",
      point_value: 15,
      receive_id: 22,
      received_name: "Scott Adams",
      send_id: 24,
      send_name: "Jim Davis"
    },
    {
      created_time: 1514939019586,
      event_id: 2,
      event_time: "2018-01-03T00:23:39.612Z",
      group_id: 1,
      note: "Thanks!",
      point_value: 10,
      receive_id: 59,
      received_name: "Bo Jackson",
      send_id: 43,
      send_name: "Deion Sanders"
    }
  ]}
]

export default mockGroupTransactions;