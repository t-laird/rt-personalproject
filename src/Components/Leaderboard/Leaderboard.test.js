/* eslint-disable max-len */

import React from 'react';
import { shallow } from 'enzyme';
import { Leaderboard, mapStateToProps } from './Leaderboard';
import mockGroupTransactions from '../../mockData/mockGroupTransactions';

describe('leaderboard tests', () => {
  it('should match the snapshot', () => {
    const renderedApp = shallow(<Leaderboard GroupTransactions={{}} />);

    expect(renderedApp).toMatchSnapshot();
  });

  it('with send_name passed in, getData should return current leaders in sending points', () => {
    const renderedApp = shallow(<Leaderboard GroupTransactions={mockGroupTransactions} />);
    const expected = [ 
      { name: 'Jim Davis', points: 15 },
      { name: 'Deion Sanders', points: 10 }
    ];

    expect(renderedApp.instance().getData('send_name')).toEqual(expected);
  });

  it('with received_name passed in, getData should return current leaders in receiving points', () => {
    const renderedApp = shallow(<Leaderboard GroupTransactions={mockGroupTransactions} />);
    const expected = [
      { name: 'Scott Adams', points: 15 }, 
      { name: 'Bo Jackson', points: 10 }
    ];

    expect(renderedApp.instance().getData('received_name')).toEqual(expected);
  });
});

describe('mapStateToProps tests', () => {
  it('should pull user transactions from store', () => {
    const mockStore = { mockGroupTransactions };
    const result = mapStateToProps(mockStore);

    expect(result.GroupTransactions).toEqual(mockStore.GroupTransactions);
  });
});