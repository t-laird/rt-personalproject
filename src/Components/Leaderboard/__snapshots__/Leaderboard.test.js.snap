// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leaderboard tests should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Leaderboard
    GroupTransactions={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          <h2>
                    Top Givers this week:
          </h2>
</div>,
        <div>
          <h2>
                    Top Earners this week:
          </h2>
</div>,
      ],
      "className": "leaderboard-component",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              Top Givers this week:
</h2>,
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Top Givers this week:",
            },
            "ref": null,
            "rendered": "Top Givers this week:",
            "type": "h2",
          },
          undefined,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              Top Earners this week:
</h2>,
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Top Earners this week:",
            },
            "ref": null,
            "rendered": "Top Earners this week:",
            "type": "h2",
          },
          undefined,
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            <h2>
                        Top Givers this week:
            </h2>
</div>,
          <div>
            <h2>
                        Top Earners this week:
            </h2>
</div>,
        ],
        "className": "leaderboard-component",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                Top Givers this week:
</h2>,
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Top Givers this week:",
              },
              "ref": null,
              "rendered": "Top Givers this week:",
              "type": "h2",
            },
            undefined,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                Top Earners this week:
</h2>,
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Top Earners this week:",
              },
              "ref": null,
              "rendered": "Top Earners this week:",
              "type": "h2",
            },
            undefined,
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
